### PR TITLE
Constant folding

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
Add functionality for constant folding and replacing nodes within the AST.

### How I did it
* Add `evaluate` method to some Vyper node classes. This method returns a new node representing the literal value, or raises `InvalidType`.
* Add `Module.replace_in_tree` for replacing an existing node with a new one.
* Add `ast/folding.py` for high-level folding methods.

### How to verify it
Run the tests. I've added additional cases using `hypothesis` to verify that the folding outcome is the same as the EVM output.

### Cute Animal Picture
image

### CODE REVIEW

File: tests/ast/nodes/test_get_children.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (fubuloubu) [inline tests/ast/nodes/test_get_children.py]: Is it important to test this API? Won't it be apparent that this doesn't work if we get other failures?

File: tests/ast/nodes/test_get_descendants.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+from pathlib import (
```
Comment (fubuloubu) [inline tests/ast/nodes/test_get_descendants.py]: Ditto this. Why test it?

File: tests/ast/nodes/test_replace_in_tree.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
+import pytest
```
Comment (fubuloubu) [inline tests/ast/nodes/test_replace_in_tree.py]: This functionality is important to test.

File: vyper/ast/folding.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    ---------
+    vyper_ast_node : Module
+        Top-level Vyper AST node.
+    """
+    while True:
+        changed_nodes = 0
+        changed_nodes += replace_literal_ops(vyper_ast_node)
+        changed_nodes += replace_subscripts(vyper_ast_node)
+        if not changed_nodes:
+            return
```
Comment (fubuloubu) [inline vyper/ast/folding.py]: Can we change this to a conditional `while` loop?
```suggestion
    changed_nodes = 42  # non-zero value to start the loop
    while changed_nodes > 0:
        changed_nodes = 0
        changed_nodes += replace_literal_ops(vyper_ast_node)
        changed_nodes += replace_subscripts(vyper_ast_node)
```

File: vyper/ast/folding.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    return changed_nodes
+
+
+def _replace(old_node, new_node):
+    if isinstance(new_node, vy_ast.Constant):
+        return new_node.from_node(old_node, value=new_node.value)
+    elif isinstance(new_node, vy_ast.List):
+        elts = [_replace(old_node, i) for i in new_node.elts]
+        return new_node.from_node(old_node, elts=elts)
```
Comment (fubuloubu) [inline vyper/ast/folding.py]: Whats an `elt`? Is it some type of deer?

File: tests/ast/nodes/test_evaluate_boolop.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    strategies as st,
+)
+
+from vyper import (
+    ast as vy_ast,
+)
+
+
+@settings(deadline=500)
+@given(a=st.booleans(), b=st.booleans(), c=st.booleans())
```
[the remark this answers is not in the document]
Comment (iamdefinitelyahuman) [reply tests/ast/nodes/test_evaluate_boolop.py]: we have `bitwise_xor`, but folding on builtin functions will be a different PR.

File: vyper/ast/folding.py (hunk 4 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/folding.py]: Yes, that looks much better

File: tests/ast/nodes/test_get_children.py (hunk 1 above)
Comment (iamdefinitelyahuman) [reply tests/ast/nodes/test_get_children.py]: Testing this behavior simplifies debugging of functionality that's reliant on it.  If a folding test fails but these tests don't, it's an area that we know not to look first.

File: tests/ast/nodes/test_get_descendants.py (hunk 2 above)
Comment (iamdefinitelyahuman) [reply tests/ast/nodes/test_get_descendants.py]: In returning descendants, the sequence is important. These tests confirm that nodes are always returned sorted by source offset and depth.

**Comment by iamdefinitelyahuman** [comment]
@fubuloubu :pray:  for showing me that I can combine pytest `parametrize` with hypothesis...  that did a lot to make the test cases more robust while reducing duplicated code.

I hadn't included nested cases with e.g. `1 + 2 + 3` because this produces an AST of `BinOp(left=1, right=BinOp(left=2, right=3))` and the `evaluate` functionality is only targeting a single node at a time. But it's a very good test..  in the commits I just pushed I've added a final case for each type of op, where I use the more high-level `folding` module to fold these nested cases, and confirm the output against the EVM when unfolded.  I like how it turned out.

File: tests/ast/nodes/test_evaluate_binop_decimal.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        places=10,
+    ),
+    right=st.decimals(
+        min_value=-(2 ** 32),
+        max_value=2 ** 32,
+        allow_nan=False,
+        allow_infinity=False,
+        places=10,
+    ),
+)
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_decimal.py]: Only write it once

```suggestion
decimals_st = st.decimals(
    min_value=-(2 ** 32),
    max_value=2 ** 32,
    allow_nan=False,
    allow_infinity=False,
    places=10,
)


@pytest.mark.fuzzing
@settings(deadline=500)
@given(left=decimals_st, right=decimals_st)
```

File: tests/ast/nodes/test_evaluate_binop_decimal.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        st.decimals(
+            min_value=-(2 ** 32),
+            max_value=2 ** 32,
+            allow_nan=False,
+            allow_infinity=False,
+            places=10,
+        ),
+        min_size=2,
+        max_size=10,
+    ),
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_decimal.py]:
```suggestion
    values=st.lists(decimals_st, min_size=2, max_size=10),
```

File: tests/ast/nodes/test_evaluate_binop_decimal.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        vy_ast.folding.replace_literal_ops(vyper_ast)
+        expected = vyper_ast.body[0].value.value
+        is_valid = True
+    except ZeroDivisionException:
+        is_valid = False
+
+    if is_valid:
+        assert contract.foo(*values) == expected
+    else:
+        assert_tx_failed(lambda: contract.foo(*values))
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_decimal.py]: Nice. This took a while to understand though, just add a comment describing what's going on here.

File: tests/ast/nodes/test_evaluate_binop_int.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    ZeroDivisionException,
+)
+
+
+@pytest.mark.fuzzing
+@settings(deadline=500)
+@given(
+    left=st.integers(min_value=-(2 ** 32), max_value=2 ** 32),
+    right=st.integers(min_value=-(2 ** 32), max_value=2 ** 32),
+)
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_int.py]: `uint64_st` for below

```suggestion
int32_st = st.integers(min_value=-(2 ** 32), max_value=2 ** 32)
uint64_st = st.integers(min_value=0, max_value=2 ** 64)


@pytest.mark.fuzzing
@settings(deadline=500)
@given(left=int32_st, right=int32_st)
```

File: tests/ast/nodes/test_evaluate_binop_int.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    else:
+        assert_tx_failed(lambda: contract.foo(left, right))
+
+
+@pytest.mark.fuzzing
+@settings(deadline=500)
+@given(
+    left=st.integers(min_value=0, max_value=2 ** 64),
+    right=st.integers(min_value=0, max_value=2 ** 64),
+)
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_int.py]: `uint64_st` from above

```suggestion
@given(left=uint64_st, right=uint64_st)
```

File: tests/ast/nodes/test_evaluate_binop_int.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+def test_binop_nested(get_contract, assert_tx_failed, values, ops):
+
+    variables = "abcdefghij"
+
+    return_value = " ".join(f"{a} {b}" for a, b in zip(variables[: len(values)], ops))
+    return_value = return_value.rsplit(maxsplit=1)[0]
+
+    source = f"""
+@public
+def foo({','.join(f'{i}: int128' for i in variables[:len(values)])}) -> int128:
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_int.py]: Do the `join` outside of the source string

File: tests/ast/nodes/test_evaluate_boolop.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+def test_boolop_nested(get_contract, values, comparators):
+
+    return_value = " ".join(
+        f"{a} {b}" for a, b in zip(variables[: len(values)], comparators)
+    )
+    return_value = return_value.rsplit(maxsplit=1)[0]
+
+    source = f"""
+@public
+def foo({','.join(f'{i}: bool' for i in variables[:len(values)])}) -> bool:
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_boolop.py]: Do the join outside of the source str

File: tests/ast/nodes/test_evaluate_boolop.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    return {return_value}
+    """
+    contract = get_contract(source)
+
+    literal_op = " ".join(f"{a} {b}" for a, b in zip(values, comparators))
+    literal_op = literal_op.rsplit(maxsplit=1)[0]
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_boolop.py]: what is this doing?

File: vyper/ast/folding.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+    return changed_nodes
+
+
+def _replace(old_node, new_node):
+    if isinstance(new_node, vy_ast.Constant):
+        return new_node.from_node(old_node, value=new_node.value)
+    elif isinstance(new_node, vy_ast.List):
+        list_values = [_replace(old_node, i) for i in new_node.elts]
+        return new_node.from_node(old_node, elts=list_values)
```
Comment (fubuloubu) [inline vyper/ast/folding.py]: Seriously, what the hell is an `elt`? Can we pick a different name?

Comment (iamdefinitelyahuman) [reply vyper/ast/folding.py]: Short for elements. It's a field name we're inheriting from python that holds all the values within a list or tuple.

We could rename it for sure, but probably belongs in a different PR.

File: tests/ast/nodes/test_evaluate_boolop.py (hunk 14 above)
Comment (iamdefinitelyahuman) [reply tests/ast/nodes/test_evaluate_boolop.py]: line 60 produces e.g. `"True and False or True and"`

line 61 trims the final `" and"`

File: tests/ast/nodes/test_evaluate_binop_int.py (hunk 12 above)
Comment (iamdefinitelyahuman) [reply tests/ast/nodes/test_evaluate_binop_int.py]: Oops! Thought I changed this in all of them, guess not.

File: tests/ast/nodes/test_evaluate_boolop.py (hunk 14 above)
Comment (fubuloubu) [reply tests/ast/nodes/test_evaluate_boolop.py]: okay gotcha

File: tests/ast/nodes/test_evaluate_binop_int.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    ZeroDivisionException,
+)
+
+st_int32 = st.integers(min_value=-(2 ** 32), max_value=2 ** 32)
+
+
+@pytest.mark.fuzzing
+@settings(max_examples=50, deadline=1000)
+@given(left=st_int32, right=st_int32)
+@pytest.mark.parametrize("op", "+-*/%")
```
Comment (fubuloubu) [inline tests/ast/nodes/test_evaluate_binop_int.py]: Add some corner cases
```suggestion
st_int32 = st.integers(min_value=-(2 ** 32), max_value=2 ** 32)


@pytest.mark.fuzzing
@settings(max_examples=50, deadline=1000)
@given(left=st_int32, right=st_int32)
@example(left=1, right=1)
@example(left=1, right=-1)
@example(left=-1, right=1)
@example(left=-1, right=-1)
@pytest.mark.parametrize("op", "+-*/%")
```
